feat: Add /mcp refresh slash command

rmedranollamas opened the pull request:
Adds a new slash command `/mcp refresh` to allow users to reload MCP servers without restarting the Gemini CLI.

This change includes:
- A new `refeshMcpServers` method in the `Config` class to refresh the tool registry.
- A new `refreshMcpAction` function in `slashCommandProcessor.ts` to handle the command logic.
- Registration of the `/mcp refresh` command and updated description for the `/mcp` command.


## Testing Matrix

<!-- Before submitting please validate your changes on as many of these options as possible -->

|          | 🍏  | 🪟  | 🐧  |
| -------- | --- | --- | --- |
| npm run  | ✔️  | ❓  | ❓  |
| npx      | ✔️  | ❓  | ❓  |
| Docker   | ❓  | ❓  | ❓  |
| Podman   | ❓  | -   | -   |
| Seatbelt | ❓  | -   | -   |

## Linked issues / bugs

- Closes #3528

rmedranollamas commented:
wait a second, I did break one thing while resolving the merge conflict. let me re-do it.

jackwotherspoon commented:
We use the term `refresh` for memory etc. Should we remain consistent and also use `refresh` for MCP?

rmedranollamas commented on packages/core/src/config/config.ts:
```diff
@@ -452,6 +452,12 @@ export class Config {
+  async reloadMcpServers(): Promise<void> {
+    if (this.toolRegistry) {
+      await this.toolRegistry.discoverTools();
+    }
```
updated, thanks for the hint, Gemi!

rmedranollamas commented:
addressed comments on the last commit in the branch. is now /mcp refresh and the reload logic is more robust.

example run:

```
(g:add-mcp-reload-command)` gemini-cli $ npm start

> @google/gemini-cli@0.1.9 start
> node scripts/start.js

Checking build status...
Build is up-to-date.

 ███            █████████  ██████████ ██████   ██████ █████ ██████   █████ █████
░░░███         ███░░░░░███░░███░░░░░█░░██████ ██████ ░░███ ░░██████ ░░███ ░░███
  ░░░███      ███     ░░░  ░███  █ ░  ░███░█████░███  ░███  ░███░███ ░███  ░███
    ░░░███   ░███          ░██████    ░███░░███ ░███  ░███  ░███░░███░███  ░███
     ███░    ░███    █████ ░███░░█    ░███ ░░░  ░███  ░███  ░███ ░░██████  ░███
   ███░      ░░███  ░░███  ░███ ░   █ ░███      ░███  ░███  ░███  ░░█████  ░███
 ███░         ░░█████████  ██████████ █████     █████ █████ █████  ░░█████ █████
░░░            ░░░░░░░░░  ░░░░░░░░░░ ░░░░░     ░░░░░ ░░░░░ ░░░░░    ░░░░░ ░░░░░


Tips for getting started:
1. Ask questions, edit files, or run commands.
2. Be specific for the best results.
3. /help for more information.


╭──────────────────╮
│  > /mcp refresh  │
╰──────────────────╯


ℹ Reloading MCP servers...


ℹ MCP servers reloaded successfully.
```

jackwotherspoon commented on packages/cli/src/ui/hooks/slashCommandProcessor.ts:
```diff
@@ -153,6 +154,41 @@ export const useSlashCommandProcessor = (
     [addItem],
   );
 
+  const reloadMcpAction = useCallback(async () => {
```
nit: should we update all `reload` usage to `refresh` to be consistent with command itself.
```suggestion
  const refreshMcpAction = useCallback(async () => {
```

jackwotherspoon commented on packages/cli/src/ui/hooks/slashCommandProcessor.ts:
```diff
@@ -153,6 +154,41 @@ export const useSlashCommandProcessor = (
+    }
+    addMessage({
+      type: MessageType.INFO,
+      content: 'Reloading MCP servers...',
```
nit: use refresh over reload..

```suggestion
      content: 'Refreshing MCP servers...',
```

jackwotherspoon reviewed (changes requested):
@rmedranollamas added a few nits to also update all functions and messages to use "refresh" over "reload".

Can you also update PR title and description too 😄

rmedranollamas commented:
all updated. thanks for those changes I missed!

jackwotherspoon commented:
@rmedranollamas  This is not working for me... `/mcp refresh` is exiting Gemini CLI for me.

Looks like lint job is failing too. May need to run `npm run preflight` again.

jackwotherspoon commented:
This is also probably unrelated to this change but another follow-up enhancement.

Typing `/memory` shows all sub-commands in the CLI... yet same behavior is not present for `/mcp`

<img width="1088" height="292" alt="image" src="https://github.com/user-attachments/assets/1ede297e-2628-4e88-ae41-d1532d975f64" />

rmedranollamas commented:
fixed the breakage (I had forgotten one renaming).

the second improved requires refactoring to a new slash command file (like memory), so I will open a new PR for that.

fijimunkii commented:
Nice one! Here's a related PR to add `/mcp off server-name` and `/mcp on server-name` https://github.com/google-gemini/gemini-cli/pull/4272
